framework/stats/vulkan_stats_provider.h

```diff
@@ -0,0 +1,163 @@
+		{
+			if (divisor_name != "")
+				scaling = StatScaling::ByCounter;
+		}
+
+		void set_vendor_graph_data(const StatGraphData &data)
+		{
+			has_vendor_graph_data = true;
+			graph_data            = data;
+		}
+
+		std::string   name;
+		StatScaling   scaling = StatScaling::ByDeltaTime;
+		std::string   divisor_name;
+		bool          has_vendor_graph_data = false;
+		StatGraphData graph_data;
+	};
+
+	using StatDataMap   = std::unordered_map<StatIndex, StatData, StatIndexHash>;
+	using VendorStatMap = std::unordered_map<StatIndex, VendorStat, StatIndexHash>;
+
+  public:
+	/**
+	 * @brief Constructs a VulkanStatsProvider
+	 * @param device The device on which to collect stats
+	 * @param requested_stats Set of stats to be collected. Supported stats will be removed from the set.
+	 * @param sampling_config Sampling mode configuration (polling or continuous)
+	 * @param num_framebuffers The number of buffers in the swapchain
+	 */
+	VulkanStatsProvider(Device &device, std::set<StatIndex> &requested_stats,
+	                    const CounterSamplingConfig &sampling_config, size_t num_framebuffers);
+
+	/**
+	 * @brief Destructs a VulkanStatsProvider
+	 */
+	~VulkanStatsProvider();
+
+	/**
+	 * @brief Checks if this provider can supply the given enabled stat
+	 * @param index The stat index
+	 * @return True if the stat is available, false otherwise
+	 */
+	bool is_available(StatIndex index) const override;
+
+	/**
+	 * @brief Retrieve graphing data for the given enabled stat
+	 * @param index The stat index
+	 */
+	const StatGraphData &get_graph_data(StatIndex index) const override;
+
+	/**
+	 * @brief Retrieve a new sample set from polled sampling
+	 * @param delta_time Time since last sample
+	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	Sample sample(float delta_time, uint32_t active_frame_idx) override;
+
+	/**
+	 * @brief A command buffer that we want stats about has just begun
+	 * @param cb The command buffer
+ 	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	void command_buffer_begun(CommandBuffer &cb, uint32_t active_frame_idx) override;
+
+	/**
+	 * @brief A command buffer that we want stats about is about to be ended
+	 * @param cb The command buffer
+	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	void command_buffer_ending(CommandBuffer &cb, uint32_t active_frame_idx) override;
+
+  private:
+	bool is_supported(const CounterSamplingConfig &sampling_config) const;
+
+	bool fill_vendor_data();
+
+	bool create_query_pools(uint32_t num_framebuffers, uint32_t queue_family_index);
+
+	float get_best_delta_time(float sw_delta_time, uint32_t active_frame_idx) const;
+
+  private:
+	Device &device;
+
+	// The query pool for the performance queries
+	std::unique_ptr<QueryPool> query_pool;
+
+	// Do we support timestamp queries
+	bool has_timestamps{false};
+
+	// The timestamp period
+	float timestamp_period{1.0f};
+
+	// And one for timestamps
```

Comment:
Comment needs to be descriptive on its own

Reply:
Done